### ISSUE TITLE
Suddenly stopped working with x11vnc on Ubuntu

### ISSUE THREAD

**Issue by Z3TA** [novnc/noVNC#1434]
This might be far fetched, posting here as a last resort.

I'm running Ubuntu 18.04.4 LTS and x11vnc to access the desktop from the web using noVNC.
A few weeks ago it suddenly stopped working... And I can't figure out why. It also stopped working on another computer about the same time, which also runs Ubuntu 18, so I think it's related to a Ubuntu unattended upgrade. Also before it stopped working, it did not work on Ubuntu 20.

Asking here hoping that someone else has more clues... Ubuntu + x11vnc + noVNC should be a fairly common combination.

How to reproduce: 
* Run Ubuntu 18 LTS
* Start x11vnc: x11vnc -rfbport 5901 -display :0 -forever
* Connect to port 5901 from noVNC (not using the proxy)

Expected: noVNC to connect successfully

Here's the logs from x11vnc and noVNC when trying to connect from older version of noVNC:
```
07/07/2020 23:28:18   other clients:
07/07/2020 23:28:18 Got 'ws' WebSockets handshake
07/07/2020 23:28:18 Got protocol: binary
07/07/2020 23:28:18   - webSocketsHandshake: using binary/raw encoding
07/07/2020 23:28:18   - WebSockets client version hybi-13
07/07/2020 23:28:18 cleaned up wsctx
07/07/2020 23:28:18 renaming client->host from '::1' to '::1'
07/07/2020 23:28:18 Disabled X server key autorepeat.
07/07/2020 23:28:18   to force back on run: 'xset r on' (3 times)
07/07/2020 23:28:18 incr accepted_client=1 for ::1:51170  sock=12
07/07/2020 23:28:18 ipv6 client: old client name: '::1'
07/07/2020 23:28:18 ipv6 client: ::1
07/07/2020 23:28:18 webSocketsDecodeHybi_enter: len=12; CTX: readlen=0 readPos=0x55f838cda830 writeTo=(nil) state=0 toRead=0 remaining=0  nReadRaw=0 carrylen=0 carryBuf=0x55f838cee0e8
07/07/2020 23:28:18 header_read to 0x55f838cda830 with len=14
07/07/2020 23:28:18 read 14 bytes from socket
07/07/2020 23:28:18 first header bytes received; opcode=2 lenbyte=12
07/07/2020 23:28:18 header complete: state=0 flen=18 writeTo=0x55f838cda83e
07/07/2020 23:28:18 calling read with buf=0x55f838cda83e and len=4 (decodebuf=0x55f838cda830 headerLen=6
)07/07/2020 23:28:18 read 4 bytes from socket; nRead=18
07/07/2020 23:28:18 toDecode=12 from n=4 carrylen=0 headerLen=6
07/07/2020 23:28:18 mask decoding; i=3 toDecode=12
07/07/2020 23:28:18 carrying over 0 bytes from 0x55f838cda84e to 0x55f838cee0e8
07/07/2020 23:28:18 copy to 12 bytes to dst buffer; readPos=0x55f838cda836, readLen=12
07/07/2020 23:28:18 after copy: readPos=(nil), readLen=0
07/07/2020 23:28:18 frame received successfully, cleaning up: read=0 hlen=6 plen=12
07/07/2020 23:28:18 cleaned up wsctx
07/07/2020 23:28:18 webSocketsDecodeHybi_exit: len=12; CTX: readlen=0 readPos=0x55f838cda830 writePos=(nil) state=0 toRead=0 remaining=0 nRead=0 carrylen=0 carryBuf=0x55f838cee0e8 result=12
07/07/2020 23:28:18 Client Protocol Version 3.8
07/07/2020 23:28:18 Protocol version sent 3.8, using 3.8
07/07/2020 23:28:18 webSocketsDecodeHybi_enter: len=1; CTX: readlen=0 readPos=0x55f838cda830 writeTo=(nil) state=0 toRead=0 remaining=0  nReadRaw=0 carrylen=0 carryBuf=0x55f838cee0e8
07/07/2020 23:28:18 header_read to 0x55f838cda830 with len=14
07/07/2020 23:28:18 read 7 bytes from socket
07/07/2020 23:28:18 first header bytes received; opcode=2 lenbyte=1
07/07/2020 23:28:18 header complete: state=0 flen=7 writeTo=0x55f838cda837
07/07/2020 23:28:18 calling read with buf=0x55f838cda837 and len=0 (decodebuf=0x55f838cda830 headerLen=6
)07/07/2020 23:28:18 toDecode=1 from n=0 carrylen=0 headerLen=6
07/07/2020 23:28:18 mask decoding; i=0 toDecode=1
07/07/2020 23:28:18 carrying over 1 bytes from 0x55f838cda837 to 0x55f838cee0e8
07/07/2020 23:28:18 cleaned up wsctx
07/07/2020 23:28:18 webSocketsDecodeHybi_exit: len=1; CTX: readlen=0 readPos=0x55f838cda830 writePos=(nil) state=0 toRead=0 remaining=0 nRead=0 carrylen=0 carryBuf=0x55f838cee0e8 result=-1
```

```
>> RFB.constructor
display.js:42 >> Display.constructor
display.js:70 User Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/83.0.4103.61 Chrome/83.0.4103.61 Safari/537.36
display.js:153 Setting viewport to full display region
display.js:92 << Display.constructor
devices.js:310 >> Keyboard.allKeysUp
devices.js:315 << Keyboard.allKeysUp
rfb.js:259 Using native WebSockets, render mode: canvas rendering
rfb.js:261 << RFB.constructor
rfb.js:536 New state 'connecting', was ''.
rfb.js:385 >> RFB.connect
rfb.js:396 connecting to ws://localhost:5901//
rfb.js:409 << RFB.connect
websock.js:222 >> WebSock.onopen
websock.js:224 Server choose sub-protocol: binary
rfb.js:213 Starting VNC handshake
websock.js:228 << WebSock.onopen
rfb.js:735 Server ProtocolVersion: 003.008
rfb.js:776 Sent ProtocolVersion: 003.008
rfb.js:806 Server security types: 1
rfb.js:831 Authenticating using scheme: 1
websock.js:231 >> WebSock.onclose
rfb.js:219 WebSocket on-close event
(anonymous) @ rfb.js:219
(anonymous) @ websock.js:232
rfb.js:592 Failed when connecting: Failed to connect to server ( (code: 1006))
```

Here's the logs when trying to connect using the latest (master branch) of noVNC:
```
07/07/2020 23:32:09 Got connection from client 127.0.0.1
07/07/2020 23:32:09   other clients:
07/07/2020 23:32:09 Got 'ws' WebSockets handshake
07/07/2020 23:32:09   - webSocketsHandshake: using base64 encoding
07/07/2020 23:32:09   - WebSockets client version hybi-13
07/07/2020 23:32:09 cleaned up wsctx
07/07/2020 23:32:09 Disabled X server key autorepeat.
07/07/2020 23:32:09   to force back on run: 'xset r on' (3 times)
07/07/2020 23:32:09 incr accepted_client=1 for 127.0.0.1:37790  sock=12
07/07/2020 23:32:09 created   xdamage object: 0x5c00040
07/07/2020 23:32:09 copy_tiles: allocating first_line at size 61
07/07/2020 23:32:18 created selwin: 0x5c00041
07/07/2020 23:32:18 called initialize_xfixes()
07/07/2020 23:32:25 selection_send: no send: uninitialized clients
07/07/2020 23:32:26 selection_send: no send: uninitialized clients
07/07/2020 23:32:49 cursor_noshape_updates_clients: 1
07/07/2020 23:32:51 selection_send: no send: uninitialized clients
07/07/2020 23:32:52 selection_send: no send: uninitialized clients
07/07/2020 23:32:57 selection_send: no send: uninitialized clients
07/07/2020 23:35:31 selection_send: no send: uninitialized clients
07/07/2020 23:35:56 selection_send: no send: uninitialized clients
07/07/2020 23:36:01 selection_send: no send: uninitialized clients
07/07/2020 23:36:09 selection_send: no send: uninitialized clients
07/07/2020 23:36:38 selection_send: no send: uninitialized clients
```
```
>> RFB.constructor rfb.js:177:13
>> Display.constructor display.js:31:13
User Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:78.0) Gecko/20100101 Firefox/78.0 display.js:61:13
<< Display.constructor display.js:69:13
<< RFB.constructor rfb.js:288:13
New state 'connecting', was ''. rfb.js:760:13
>> RFB.connect rfb.js:505:13
connecting to ws://127.0.0.1:5901// rfb.js:507:13
<< RFB.connect rfb.js:554:13
GETws://127.0.0.1:5901//
[HTTP/1.1 101 Switching Protocols 40ms]

>> WebSock.onopen websock.js:190:17
Starting VNC handshake rfb.js:242:21
<< WebSock.onopen websock.js:196:17
Ignoring empty message websock.js:291:17
```

**Comment by samhed** [comment]
You're having problems with both the old noVNC (how old?) and the latest master as well?

Could you elaborate on what happens when it "doesn't work"? The logs are not particularly clear here.

**Comment by Z3TA** [comment]
It just says "connecting..." then it gives up after a few seconds.

The version in package.json is 0.6.1 for the "old" version of noVNC I tested. And 1.2.0-beta for master branch.

This is likely not an issue with noVNC, I believe a library that libvncserver depends on got an update in Ubuntu which somehow broke the protocol.

**Comment by CendioOssman** [comment]
I see this issue:

> 07/07/2020 23:32:09   - webSocketsHandshake: using base64 encoding

That indicates that your x11vnc is too old for modern noVNC. Could you find something more current and see if that works better? I'm afraid we cannot help you out with the old 0.6.1 version of noVNC.

**Comment by Z3TA** [comment]
I tried with latest x11vnc although I did still have the Ubuntu libs for libvncserver where I think the problem is.

**Comment by Z3TA** [comment]
I have another computer Running Ubuntu 20 and it has a newer version of libvncserver:
$ dpkg -l | grep libvncserver
ii  libvncserver1:amd64                        0.9.12+dfsg-9ubuntu0.1                amd64        API to write one's own VNC server

I still have the same issue with the latest/master version of noVNC, but if I however checkout an earlier release of noVNC it works!

`10/07/2020 02:17:48 Got 'ws' WebSockets handshake                                                                                                                                     
10/07/2020 02:17:48 Got protocol: binary          `

I can connect using noVNC v1.1.0 but not using v1.2.0-beta or master

`10/07/2020 02:22:48 Got 'ws' WebSockets handshake                                                                                                                                     
10/07/2020 02:22:48   - webSocketsHandshake: using base64 encoding   `

**Comment by Z3TA** [comment]
To summarize the situation: Ubuntu 18 got a patch which broke protocol with all noVNC versions. Ubuntu 20 has a newer version of libvncserver which only works with older versions of noVNC.

**Comment by samhed** [comment]
libvncserver fixed their bug in version 0.9.13

**Comment by samhed** [comment]
See https://github.com/novnc/noVNC/issues/1310
